core/common/src/main/java/alluxio/resource/CloseableIterator.java

```diff
@@ -18,32 +18,67 @@
 import java.io.IOException;
 import java.util.Iterator;
 import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.function.Consumer;
 
 /**
  * A {@code CloseableIterator<T>} is an iterator which requires cleanup when it is no longer in use.
  *
+ * When the user creates {@code CloseableIterator}, the {@code closeResource()} method needs to be
+ * implemented to define how the resource can be closed.
+ * The iterator is not necessarily {@code Closeable} so we do not provide a default implementation
+ * for {@code closeResource()}.
+ *
  * @param <T> the type of elements returned by the iterator
  */
-public abstract class CloseableIterator<T> extends CloseableResource<Iterator<T>> {
+public abstract class CloseableIterator<T> extends CloseableResource<Iterator<T>> implements Iterator<T> {
+  Iterator<T> mIter;
+
   /**
    * Creates a {@link CloseableIterator} wrapper around the given iterator. This iterator will
    * be returned by the {@link CloseableIterator#get()} method.
    *
    * @param iterator the resource to wrap
    */
-  public CloseableIterator(Iterator<T> iterator) {
+  CloseableIterator(Iterator<T> iterator) {
```

Comment:
Reduce the access on the constructor, as we have the factory methods already. It's not `private` only because of one inheritance in `CloseableIteratorTest`.